Create Autometrics specification v1.0.0 (#85)
autometrics-dev/autometrics-shared: specs/autometrics_v1.0.0.md

This is a specification for Autometrics. Its high level goal is to enable
operability between the various different Autometrics libraries and consumers.

- [API](#api)
  - [Service-Level Objectives (SLOs)](#service-level-objectives-slos)
- [Conventions](#conventions)
- [Metrics](#metrics)
  - [`function.calls`](#functioncalls)
  - [`function.calls.duration`](#functioncallsduration)
  - [`build_info`](#build_info)
  - [`function.calls.concurrent`](#functioncallsconcurrent)
- [Labels](#labels)
  - [`autometrics.version`](#autometricsversion)
  - [`branch`](#branch)
  - [`caller.function`](#callerfunction)
  - [`caller.module`](#callermodule)
  - [`commit`](#commit)
  - [`function`](#function)
  - [`module`](#module)
  - [`objective.name`](#objectivename)
  - [`objective.percentile`](#objectivepercentile)
  - [`objective.latency_threshold`](#objectivelatency_threshold)
  - [`result`](#result)
  - [`service.name`](#servicename)
  - [`version`](#version)
- [Changelog](#changelog)
  - [v1.0.0](#v100)

## API

Libraries SHOULD expose a decorator, macro, wrapper function, or use another
meta-programming technique offered by the language to instrument functions and

> methods in the user's source code. Ideally, the function attribute should simply

Reviewer: ```suggestion
methods in the user's source code. Ideally, the function attribute SHOULD simply
```

be called `autometrics` or `Autometrics`, but libraries MAY append a suffix to
the name if necessary.

Libraries MAY enable the decorator, macro, etc to apply to an entire class
definition. If they do, they SHOULD provide an option for users to skip or
ignore particular methods.

Libraries MAY need an initialization function.

Libraries SHOULD expose functionality to make the metrics available to the
user in either the Prometheus text format or the OpenTelemetry Protocol (otlp).
It is up to the user to decide whether they want to implement a push or pull
based model. Libraries MAY provide higher level implementation to expose the
metrics in a push or pull based model.

### Service-Level Objectives (SLOs)

Libraries SHOULD expose functionality to create objectives within the source
code. Objectives can be "attached" to functions by passing the objective to the
Autometrics decorator, macro, etc for one or more functions.

Objectives can relate to functions' success rate and/or latencies.

> Success rate objectives add the [`objective.name`](#objectivename) and

Reviewer: i assume this is a requirement?

```suggestion
Success rate objectives MUST add the [`objective.name`](#objectivename) and
```

[`objective.percentile`](#objectivepercentile) labels to the
[`function.calls`](#functioncalls) metric.

> Latency objectives add the [`objective.name`](#objectivename),

Reviewer: same here

```suggestion
Latency objectives MUST add the [`objective.name`](#objectivename),
```

[`objective.percentile`](#objectivepercentile), and
[`objective.latency_threshold`](#objectivelatency_threshold) labels to the
[`function.calls.duration`](#functioncallsduration) metric.

## Conventions

Autometrics uses the [OpenTelemetry Metric Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md)
for naming metrics, including using `.`'s as separators.

When the metrics are exported to Prometheus, all dot (`.`) separators in the
metric name and label keys are replaced by underscores (`_`). Suffixes are
appended where required by Prometheus/OpenMetrics.

> Label values MAY contain any Unicode characters.

Author: I think so, this part came from the original spec.

## Metrics

The following is a list of metrics that MUST be exported when annotated by an
autometrics library. The only exception is the `function.calls.concurrent`
metrics, which a library MAY support.

### `function.calls`

This metric is a 64-bit monotonic counter that tracks the number of times a
given function was invoked. It stores information about the result of the
function, and optionally which function called the given function.

**Metric type:** Counter

**OTEL metric name:** `function.calls`

**Prometheus metric name:** `function_calls_total`

| Label | Optional/required | Notes |
|-------|-------------------|-------|
| [`function`](#function) | required | |

> | [`module`](#module) | required | |

Reviewer 2: When methods are instrumented, is the name of the type part of the function name or is it somehow part of the module? I think it makes sense for the specification to offer guidance on this.

Reviewer 3: I think we mostly wanted to put the class name as part of the function name [last time we discussed it](https://github.com/orgs/autometrics-dev/discussions/33). I don't have strong opinion one way or another on this, but it seemed technically easier to have this class part of the function name

Reviewer 2: Good call. I think that discussion is an important one to resolve before 1.0, so at least there's some guidance. Although with the `module` field already being freeform, maybe the urgency isn't even that high :)

To offer my two cents: I'd favor adding types to the `function` field, because module is a distinct concept in most languages. It also allows distinction in the `function` field between instance methods and static methods, for instance. For instance, JS/TS have the convention to denote instance methods with `#` while static methods use `.`.

Examples:

<table>
<tbody>
<tr>
<td></td><td><b>module</b></td><td><b>function</b></td>
</tr>
<tr>
<td>Rust</td><td><code>foo::bar::qux</code></td><td><code>Foo::bar</code></td>
</tr>
<tr>
<td>TypeScript (static method)</td><td><code>/foo/bar/qux.js</code></td><td><code>Foo.bar</code></td>
</tr>
<tr>
<td>TypeScript (instance method)</td><td><code>/foo/bar/qux.js</code></td><td><code>Foo#bar</code></td>
</tr>
</tbody>
</table>

The only downside I see is that explorer might need a little bit of extra complexity if it wants to offer things like "expanding/collapsing" entire classes or types in the tree view. But again, we're already pushing all the complexity for language-specific module formats there as well, so I don't think that's a big deal.

Reviewer 3: Also, for the record I think the Go implementation will still leave the class/struct name in the module part, because contrary to Rust, in Go it's actually a lot harder (I think it's impossible) to differentiate subpackages from class names in all cases. See [this Go playground](https://go.dev/play/p/xrdc1HRAfI8) for example, where if another instrumented codebase had a `helper` subpackage, it would be impossible to disambiguate the 2 cases.

I think this makes "having the version/language of bundled autometrics" as part of `build_info` quite important, so that all the unspecified bits have a reference to go to for these small details. I.e. "Oh `build_info` tells me it's Rust so I should split function names by `::` if I want class names separately" or something like that for the eventual consumers of the metrics

Author: That is indeed an item which we should include in this spec. I do want to emphasis that the current commits in this PR are not going to be final yet. There will still be some other discussions that need to be resolved, which might result in new commits in this PR. The final thing will eventually be merged.

Author: If we still want to get this into 1.0 I think it would be important to get a PR out for the changes. Can either @gagbo or @arendjr get on this, if we want to discuss this further, then I suggest moving this into a discussion (though there might already be a relevant discussion open)

Reviewer 3: Sure https://github.com/autometrics-dev/autometrics-shared/pull/94

> | [`module`](#module) | required | |

Reviewer 2: How are modules formatted? Do we mandate them to use a Rust-like separators such as `::` for path segments? Or can it also be a `/`-separated path? Or a `.`-separated Java-style namespace?

Author: As it currently is, there are no restriction on this at all. The only requirement is that the module+function should be unique. I think it is best to not put any requirements on the module content itself, as every language has its own conventions (as enumerated by you)

Reviewer 2: Yeah, it pushes a bit of complexity down to the explorer, but I think that may be fair indeed :+1:

|-------|-------------------|-------|
| [`function`](#function) | required | |
| [`module`](#module) | required | |
| [`service.name`](#servicename) | required | |
| [`result`](#result) | required | |
| [`caller.function`](#callerfunction) | optional | |
| [`caller.module`](#callermodule) | optional | |
| [`objective.name`](#objectivename) | optional | if an [objective](#service-level-objectives-slos) is attached to the given function |
| [`objective.percentile`](#objectivepercentile) | optional | if an [objective](#service-level-objectives-slos) is attached to the given function |

If the library is unable to determine whether a function was successful or not,

> then it MUST always set the result to `ok`.

Reviewer: shouldnt we have like a third, undetermined state, e.g `n/a`? i think it would make sense to have such a state so we can within explorer guide users to add function success/erroring? maybe even warn if they have slo's that depend on it (but which will never trigger if we cant figure out if its successful or not)?

Author: I think that could definitely be useful. Maybe we should first create a discussion for this and then we can add it to the 1.0.0 or any feature versions (depending on the outcome of the discussions).

Reviewer: done: https://github.com/orgs/autometrics-dev/discussions/95

If the library is able to determine which function called the given function,
then it MUST set the caller information in the `caller.function` and
`caller.module` labels. If the caller is not known, either it is the entrypoint
or the library is unable to determine the caller, then the `caller.function` and

> `caller.module` labels MUST be absent.

Author: So you are suggesting that we always require these labels to be set, but if they are not available then they should always be set to an empty string, `""`?

What do you think about the other ones? Like labels associated with objectives, and things in the `build_info` metric.

Also this gravel issue that you mention will become an issue once we have newer specs with new features. If there are multiple apps running with that target different specs then it might not work with gravel anymore. Tbh I think we should be able to use OTLP for most push related metrics, so it _might_ not be a big issue.

> `caller.module` labels MUST be absent.

Reviewer 3: I think we can leave the choice of having an empty value or skipping the label altogether. This choice does not seem to impact consumers of autometrics metrics too much, and that means on the implementation side we don't have to look into prometheus or otel libraries implementations to make sure that it complies to stricter specification, i.e. "What happens if I declare a metric with a `caller.function` label, but do not mention this label when I interact with the counter later?" which could be any of:
- compilation error,
- runtime error,
- empty label,
- absent label

and this disambiguation would need to happen on every use of underlying metrics and exporter libraries

### `function.calls.duration`

This metric is a 64-bit floating point histogram that tracks the duration or
latency of function calls.

**Metric type:** Histogram

**OTEL metric name:** `function.calls.duration`

**Prometheus Name:** `function_calls_duration_seconds`

| Label | Optional/required | Notes |
|-------|-------------------|-------|
| [`function`](#function) | required | |
| [`module`](#module) | required | |
| [`service.name`](#servicename) | required | |
| [`objective.name`](#objectivename) | optional | if an [objective](#service-level-objectives-slos) is attached to the given function |
| [`objective.percentile`](#objectivepercentile) | optional | if an [objective](#service-level-objectives-slos) is attached to the given function |

A library MUST track the duration in seconds.

Libraries MUST use the [default OpenTelemetry histogram buckets](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#histogram-aggregations)
as a default. The Library SHOULD allow the user to override these default buckets.

### `build_info`

A metrics that contains metadata related to the application. It should be

> possible to join this metric with other metrics to enrich other metrics.

Reviewer: maybe we should also specify what meta data, e.g `version` and git `sha`

Author: the table below lists the labels that are required, do you think that is enough? I'm not a fan of copying the data since that will cause them to drift apart at some point.

**Metric type:** Gauge

> **OTEL metric name:** `build_info`

Author: I would prefer not to introduce too many backwards incompatible changes. But I guess this is the best time to do it.

### `autometrics.version`

The version of the specification that the library targets. This version MUST
contain the full version without the `v` prefix.

Consumers MAY only use the major and minor version, as the patch version changes
SHOULD be made in a backwards compatible way.

For the current specification this SHOULD be `1.0.0`.

### `branch`

The Git branch of the user's project. If this information is not available, this
label MAY be absent or empty (`""`).

### `caller.function`

The name of the `function` that invoked the given function. If the caller is
not known, this label MUST be absent or empty (`""`).

This SHOULD refer to Autometrics-instrumented functions. Therefore, if Function
A calls Function B, which calls Function C and only Functions A and C are
instrumented but not B, the `caller` of Function C would be Function A.

Libraries MAY make this label optional (on an opt-out basis) if collecting this
information has a non-negligible performance overhead.

### `caller.module`

The `module` of the function that invoked the given function. If the caller is
not known, this label MUST be absent or empty (`""`).

See [`caller.function`](#callerfunction)

> 

Reviewer 3: We should mention that the names here must match `function` and `module` labels of the original function definition (so even if there's aliasing involved, everything points to the same combination)

Author: Isn't this covered by the paragraph on line 207?

Reviewer 3: True true

MUST contain the user-specified name of the objective. If there is no SLO
attached, this label MUST be absent or empty (`""`).

### `objective.percentile`

If a function has an [SLO](#service-level-objectives-slos) attached, this label
MUST specifies the percentage of requests that should return the `result="ok"`
OR the percentage of requests that should meet the specified
[`objective.latency_threshold`](#objectivelatency_threshold).

The value MUST be expressed as a percentage, so 99.9% would be `"99.9"`
(without the `%` symbol).

If there is no SLO attached, this label MUST be absent or empty (`""`).

Libraries SHOULD support the following percentiles: `"90"`, `"95"`, `"99"`,
`"99.9"`. Libraries MAY allow users to specify custom percentiles but care
should be taken to ensure that users generate separate Prometheus recording
rules for the custom percentiles.

### `objective.latency_threshold`

If a function has an [SLO](#service-level-objectives-slos) attached, this MUST
specify the maximum duration of function calls that are considered meeting
the objective.

This MUST be specified in seconds (**not** milliseconds).

Libraries SHOULD support the

> [default OpenTelemetry histogram buckets](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#histogram-aggregations)

Reviewer 3: We should put the values again here in case otel changes its default values or moves this page (the link here does not specify a version of OTEL spec).

Reviewer 4: agreed!

Author: > We should put the values again here in case otel changes its default values or moves this page (the link here does not specify a version of OTEL spec).

I copied over the values from the otel spec, but I'm not sure those are of any use to us. Those values are: 
```[ 0, 5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000 ]```

Since this is about latency in seconds, almost all will be in the first or second bucket.

I'm going to checkout what the current defaults are in the rust implementation now.

Author: I found this reference to the default buckets in the Rust implementation: https://github.com/autometrics-dev/autometrics-rs/blob/c9bd7e1285ccc721e16083b8cf4b2553786d65bd/autometrics/src/settings.rs#L20C1-L23C3

as label values. Libraries MAY allow users to specify custom latencies but care
should be taken to ensure that the value of this label matches one of the
histogram buckets supported by the [`function.calls.duration`](#functioncallsduration)
metric.

### `result`

Whether the function executed successfully or errored. An error MAY either mean
that the function returned an error or that it threw an exception.

The value of this label MUST either be `"ok"` or `"error"`.

Libraries MAY offer users the ability to override the default behavior for
determining whether the `result` label should be `"ok"` or `"error"`, for
example to allow users to treat client-side errors as `"ok"`.

### `service.name`

The logical name of a service. This matches the
[OpenTelemetry Service specification](https://github.com/open-telemetry/semantic-conventions/tree/main/specification/resource/semantic_conventions#service).

All metrics produced by a library from a given instance SHOULD use a single
`service.name`. All instances of a horizontally scaled service SHOULD also use
the same `service.name`.

Libraries SHOULD support setting the `service.name` using environment variables

> (`AUTOMETRICS_SERVICE_NAME` and `OTEL_SERVICE_NAME`, with the first taking

Reviewer: we shoudl define a default if these are not set, e.g the projects name used by the buîld system

Author: Ah right, I guess we can advice the a fallback for the library.

# Autometrics specification v1.0.0 <!-- omit in toc -->

This is a specification for Autometrics. Its high level goal is to enable

> operability between the various different Autometrics libraries and consumers.

Reviewer: maybe we should also mention that the keywords used in this document (`MUST`, `SHOULD` etc.) should be interpreted according to RFC 2119: https://www.rfc-editor.org/rfc/rfc2119

Author: I've added a line about this.

Libraries MAY allow users to specify custom latencies, but care
should be taken to ensure that the value of this label matches one of the
histogram buckets supported by the [`function.calls.duration`](#functioncallsduration)
metric.

### `result`

Whether the function executed successfully or errored. An error MAY either mean
that the function returned an error or that it threw an exception.

The value of this label MUST either be `"ok"` or `"error"`.

Libraries MAY offer users the ability to override the default behavior for
determining whether the `result` label should be `"ok"` or `"error"`, for
example to allow users to treat client-side errors as `"ok"`.

### `service.name`

The logical name of a service. This matches the
[OpenTelemetry Service specification](https://github.com/open-telemetry/semantic-conventions/tree/main/specification/resource/semantic_conventions#service).

All metrics produced by a library from a given instance SHOULD use a single
`service.name`. All instances of a horizontally scaled service SHOULD also use
the same `service.name`.

Libraries SHOULD support setting the `service.name` using environment variables

> (`AUTOMETRICS_SERVICE_NAME` and `OTEL_SERVICE_NAME`, with the first taking

Reviewer: maybe also specify wheter this is compile-time env variable value or runtime

Author: That is a good point, I guess that applies to all environment variables. Though I think we should consider them all to be runtime environment variables (possibly with a compile time default). Since not all languages have a compile phase...

Author: I've added this section: https://github.com/autometrics-dev/autometrics-shared/pull/85/commits/76b4681723b5b0a7b1021141104bea9c51f75eb1

**Metric type:** Gauge

**OTEL metric name:** `build_info`

**Prometheus Name:** `build_info`

| Label | Optional/required | Notes |
|-------|-------------------|-------|
| [`version`](#version) | optional | |
| [`commit`](#commit) | optional | |
| [`branch`](#branch) | optional | |
| [`service.name`](#servicename) | required | |
| [`autometrics.version`](#autometricsversion) | required | |
| [`repository.url`](#repositoryurl) | optional | |
| [`repository.provider`](#repositoryprovider) | optional | |

This is a gauge or up/down counter. It MUST always have the value of `1.0`.

### `function.calls.concurrent`

An optional metric that tracks the number of concurrent calls to a given
function.

**Metric type:** Gauge

**OTEL metric name:** `function.calls.concurrent`

**Prometheus Name:** `function_calls_concurrent`

| Label | Optional/required | Notes |
|-------|-------------------|-------|
| [`function`](#function) | required | |
| [`module`](#module) | required | |
| [`service.name`](#servicename) | required | |

## Labels

The following is a list of labels and their intended use.

### `autometrics.version`

> Used with metrics: [`build_info`](#build_info)

> The version of the specification that the library targets. This version MUST
> contain the full version without the `v` prefix.
> 
> Consumers MAY only use the major and minor version, as the patch version changes
> SHOULD be made in a backwards compatible way.
> 
> For the current specification this SHOULD be `1.0.0`.

Reviewer 2: There's a slight contradiction here, because you first say the version MUST be the full version, then later say it MAY not have to be. I would actually lean towards suggesting `{major}.{minor}` by default, and only leaving open the possibility to specify a patch version if necessary.

```suggestion
The version of the specification that the library targets. This version MUST contain the
semantic version without the `v` prefix.

Consumers SHOULD only use the major and minor version, as the patch version changes
SHOULD be made in a backwards compatible way. If necessary, the patch version MAY be
specified too.

For the current specification this SHOULD be `1.0`.
```

Author: The library that is exporting the metrics should use the full version that it is targeting, so something like `1.0.0`. But a consumer (like explorer) can make a query that will use any major/minor version using a regex query, something like this: `build_info{autometrics.version=~"1.0*"}`. 

I think it is better to know exactly what version is targetted by the library from the metrics themselves, instead of needing to look it up in some docs somewhere.

Reviewer 2: Ah right, I misinterpreted what the consumers were here. My bad!

the same `service.name`.

Libraries SHOULD support setting the `service.name` using environment variables
(`AUTOMETRICS_SERVICE_NAME` and `OTEL_SERVICE_NAME`, with the first taking
precedence if both are set). Libraries MAY also support configuring this value
in an initialization function. If the service name is not set by the user, then
the library MUST set a default based on the user's project, this could be
something like a package name or the binary name.

### `version`

> Used with metrics: [`build_info`](#build_info)

The version of the user's project, ideally using
[Semantic Versioning](https://semver.org/). It SHOULD only contain the version
number and SHOULD NOT start with a `v`.

### `repository.url`

> Used with metrics: [`build_info`](#build_info)

A URL to the user's project git or other scm repository. This SHOULD be a URL
that makes sense for the repository type. For example, for a git repository, it
MAY be a HTTP URL or a SSH URL.

A library SHOULD use the value from the `AUTOMETRICS_REPOSITORY_URL` environment
variable, if set. It SHOULD also allow for a value to be specified in the
initialization function. A library MAY also attempt to determine the repository

> by itself, but the user MUST be able to opt-out of this behavior.

Reviewer 5: how would that work if for example user didn't set the init setting and `AUTOMETRICS_REPOSITORY_URL`, do we need to make a separate setting/env var to not look up other env vars like `GITHUB_REPOSITORY` in this case?

Author: I think this could be left up to the library implementation. But I would expect a envvar to be set which will make the library not do any auto discovery, maybe something like `AUTOMETRICS_NO_GITHUB_REPOSITORY=1`. 

Do you think that is sufficient or would you suggest something else?

Reviewer 2: I have interpreted this as: explicitly setting `repository.url` to an empty string in the init function is the way to opt-out of the auto-detected default.

Implemented it like that here: https://github.com/autometrics-dev/autometrics-ts/pull/138

A library MAY try to parse the `repository.url` to determine the provider. It
MUST allow the user to specify their own value using the
`AUTOMETRICS_REPOSITORY_PROVIDER` environment variable. The library MAY use the

> initialization function to override this value as well.

Reviewer 2: I find the MUST/MAY wording a little too strong here. For `autometrics-ts` environment variables cannot be used in the client-side use case, so it could be argued the client-side implementation cannot be spec-compliant this way. Maybe a wording such as this could help: "It MUST allow the user to specify their own value to set the provider. Where applicable, the `AUTOMETRICS_REPOSITORY_PROVIDER` environment variable SHOULD be accepted for this purpose. The library MAY also use the initialization function to set this value. If both are present, the value passed to the initialization function MUST take precedence.

Author: I've implemented this in #91
